File: runme/mcgen/local/datan/exasrc/e7lsq.c
#include <stdio.h>
#include "datsrc.h"

/* Table of constant values */

static integer c__8 = 8;

main(void)
{
    /* Initialized data */ /* Local variables */

    static integer i, m, n, nstep, nr, nred, list[2];
    static double t[4] = { .2,.5,.8,1.3 }, s[4] = { .15,.4,.7,.8 },
                  dt[4] = { .1,.1,.1,.05 }, ds[4] = { .05,.1,.1,.1 },
                  rho[4] = { 0.,0.,.5,0. }, e[40] /* was [4][10] */,
                  f[100] /* was [10][10] */, r, x[2], y[8],
                  a2[100] /* was [10][10] */, cx[4] /* was [2][2] */,
                  cy[64] /* was [8][8] */, gy[64] /* was [8][8] */;

    /* System generated locals */
    double d__1;

/* identify program to user */
    printf("%s\n\n", "Program E7LSQ demonstrates use of LSQGEN.");
/* write table of data */
    printf("%s\n", "    T         S         DT        DS        RHO");
    for (i = 1; i <= 4; ++i) {
        printf("%10.5lf%10.5lf%10.5lf%10.5lf%10.5lf\n", 
               t[i-1], s[i-1], dt[i-1], ds[i-1], rho[i-1]);
    }
    printf("\n");
/* set up data for input to LSQGEN */
    n = 8;
    m = 4;
    nr = 2;
    for (nred = 2; nred >= 0; --nred) {
	mtxunt_(cy, &c__8);
	for (i = 1; i <= 4; ++i) {
	    y[(i - 1) * 2] = t[i - 1];
	    y[(i - 1 << 1) + 1] = s[i - 1];
/* Computing 2nd power */
	    d__1 = dt[i - 1];
	    cy[(i - 1 << 1) + 1 + ((i - 1 << 1) + 1 << 3) - 9] = d__1 * d__1;
/* Computing 2nd power */
	    d__1 = ds[i - 1];
	    cy[(i - 1 << 1) + 2 + ((i - 1 << 1) + 2 << 3) - 9] = d__1 * d__1;
	    cy[(i - 1 << 1) + 1 + ((i - 1 << 1) + 2 << 3) - 9] = rho[i - 1] * 
		    ds[i - 1] * dt[i - 1];
	    cy[(i - 1 << 1) + 2 + ((i - 1 << 1) + 1 << 3) - 9] = rho[i - 1] * 
		    ds[i - 1] * dt[i - 1];
	}
/* determine first approximation */
	if (nred == 2) {
	    list[0] = 1;
	    list[1] = 1;
	    x[1] = (s[3] - s[0]) / (t[3] - t[0]);
	    x[0] = s[0] - x[1] * t[0];
	} else if (nred == 1) {
	    list[0] = 0;
	    list[1] = 1;
	    x[0] = .2;
	    x[1] = (s[3] - s[0]) / (t[3] - t[0]);
	} else if (nred == 0) {
	    list[0] = 0;
	    list[1] = 0;
	    x[0] = 0.;
	    x[1] = .5;
	}
/* header for output of results */
        printf("%s\n", "performing fit with LSQGEN");
        printf("%s%2li%s%2li%s%2li%s%2li%2li\n", "N = ", n, ", NR = ", nr,
               ", NRED = ", nred, ", LIST = ", list[0], list[1]);
        printf("%s%10.5lf%10.5lf\n", "first approx.: X = ", x[0], x[1]);
	nstep = 100;
	lsqgen_(y, cy, gy, f, e, &m, &n, &nr, &nred, list, x, cx, &r,
               a2, &nstep);
/* output of results */
        printf("%s%6.2lf%s%3li\n%s%10.5lf%10.5lf\n", "result of fit: R = ", r,
               ", NSTEP =", nstep, "X =", x[0], x[1]);
	if (nred > 0) {
            printf("%s\n", "covariance matrix CX = ");
	    mtxwrt_(cx, &nred, &nred);
	}
        printf("\n");
    }
    return 0;
} /* main */

File: runme/mcgen/local/datan/exasrc/e7lsq.f
      PROGRAM E7LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXN=8,MAXNR=2,MAXM=4)
      DIMENSION X(MAXNR),Y(MAXN),CY(MAXN,MAXN)
      DIMENSION GY(MAXN,MAXN),CX(MAXNR,MAXNR)
      DIMENSION F(MAXN+MAXNR,MAXN+MAXNR)
      DIMENSION E(MAXM,MAXN+MAXNR)
      DIMENSION A2(MAXN+MAXNR,MAXN+MAXNR),LIST(2)
      DIMENSION T(4),S(4),DT(4),DS(4),RHO(4)
      DATA T /0.2D0,0.5D0,0.8D0,1.3D0/
      DATA S /0.15D0,0.4D0,0.7D0,0.8D0/
      DATA DT/0.1D0,0.1D0,0.1D0,0.05D0/
      DATA DS/0.05D0,0.1D0,0.1D0,0.1D0/
      DATA RHO/0.D0,0.D0,0.5D0,0.D0/
C identify program to user
      WRITE(*,*)' Program E7LSQ demonstrates use of LSQGEN'
      WRITE(*,*)' '
C write table of data
      WRITE(*,*)'    T         S         DT        DS        RHO'
      DO 5 I=1,4
        WRITE(*,'(5F10.5)')T(I),S(I),DT(I),DS(I),RHO(I)
5     CONTINUE
      WRITE(*,'(/)')
C set up data for input to LSQGEN
      N=8
      M=4
      NR=2
      DO 20 NRED=2,0,-1
        CALL MTXUNT(CY,8)
        DO 10 I=1,4
          Y((I-1)*2+1)=T(I)
          Y((I-1)*2+2)=S(I)
          CY((I-1)*2+1,(I-1)*2+1)=DT(I)**2
          CY((I-1)*2+2,(I-1)*2+2)=DS(I)**2
          CY((I-1)*2+1,(I-1)*2+2)=RHO(I)*DS(I)*DT(I)
          CY((I-1)*2+2,(I-1)*2+1)=RHO(I)*DS(I)*DT(I)
10      CONTINUE
C determine first approximation
        IF(NRED.EQ.2) THEN
          LIST(1)=1
          LIST(2)=1
          X(2)=(S(4)-S(1))/(T(4)-T(1))
          X(1)=S(1)-X(2)*T(1)
        ELSE IF(NRED.EQ.1) THEN
          LIST(1)=0
          LIST(2)=1
          X(1)=.2D0
          X(2)=(S(4)-S(1))/(T(4)-T(1))
        ELSE IF(NRED.EQ.0) THEN
          LIST(1)=0
          LIST(2)=0
          X(1)=0.D0
          X(2)=0.5D0
        END IF
C header for output of results
        WRITE(*,*)' performing fit with LSQGEN'
        WRITE(*,'(3(A,I2),A,2I2)')' N = ',N,', NR = ',NR,
     +  ', NRED = ',NRED,', LIST = ',LIST
        WRITE(*,'(A,2F10.5)')' first approx.: X = ',X
        NSTEP=100
        CALL LSQGEN(Y,CY,GY,F,E,M,N,NR,NRED,LIST,X,CX,R,A2,NSTEP)
C output of results
        WRITE(*,'(A,F6.2,A,I3,/,A,(2F10.5))')
     +  ' result of fit: R = ',R,', NSTEP =',NSTEP,' X =',X
        IF(NRED.GT.0) THEN
          WRITE(*,*)' covariance matrix CX = '
          CALL MTXWRT(CX,NRED,NRED)
        END IF
        WRITE(*,'(/)')
20    CONTINUE
      END
